### ISSUE TITLE
torch.cuda.make_graphed_callables can silently corrupt parameter gradient accumulation by returning static grad buffers

### ISSUE THREAD

**Issue by buptzyb** [pytorch/pytorch#181723]
# Title

`torch.cuda.make_graphed_callables` can silently corrupt parameter gradient accumulation by returning static grad buffers

# Body

### Bug description

`torch.cuda.make_graphed_callables` can silently produce incorrect parameter gradients when a graphed module is used with gradient accumulation over multiple microbatches.

The backward autograd function returns detached views of captured `static_grad_inputs`. For module parameters, those returned tensors flow into `AccumulateGrad`. If `param.grad is None`, `AccumulateGrad` may install/steal the incoming grad tensor as `param.grad`. In that case, `param.grad` aliases the CUDA graph's static parameter grad buffer. The next backward graph replay overwrites the same static buffer before `AccumulateGrad` accumulates the next microbatch's gradient.

This makes the accumulated gradient become `g2 + g2` instead of `g1 + g2`.

This is silent numerical corruption.

### Minimal repro

```python
import torch


class ScaleByWeight(torch.nn.Module):
    def __init__(self, width: int) -> None:
        super().__init__()
        self.weight = torch.nn.Parameter(
            torch.arange(1, width + 1, device="cuda", dtype=torch.float32)
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x * self.weight


torch.cuda.set_device(0)
torch.manual_seed(1234)

width = 4
model = ScaleByWeight(width).cuda()
sample_x = torch.ones(2, width, device="cuda", dtype=torch.float32)
graphed_model = torch.cuda.make_graphed_callables(model, (sample_x,))

x1 = torch.tensor(
    [[1.0, 2.0, 3.0, 4.0], [2.0, 3.0, 4.0, 5.0]],
    device="cuda",
)
x2 = torch.tensor(
    [[10.0, 20.0, 30.0, 40.0], [1.0, 2.0, 3.0, 4.0]],
    device="cuda",
)

expected_g1 = x1.sum(dim=0)
expected_g2 = x2.sum(dim=0)
expected_total = expected_g1 + expected_g2
bug_total = expected_g2 + expected_g2

model.weight.grad = None

graphed_model(x1).sum().backward()
torch.cuda.synchronize()

first_grad_ref = model.weight.grad
first_grad_ptr = first_grad_ref.data_ptr()
first_grad_snapshot = first_grad_ref.detach().clone()

observations = {}


def capture_pre_accumulate(incoming_grad: torch.Tensor) -> torch.Tensor:
    existing_grad = model.weight.grad
    observations["incoming_ptr"] = incoming_grad.data_ptr()
    observations["incoming_value"] = incoming_grad.detach().clone()
    observations["existing_ptr"] = None if existing_grad is None else existing_grad.data_ptr()
    observations["existing_value"] = (
        None if existing_grad is None else existing_grad.detach().clone()
    )
    return incoming_grad


hook_handle = model.weight.register_hook(capture_pre_accumulate)
try:
    graphed_model(x2).sum().backward()
    torch.cuda.synchronize()
finally:
    hook_handle.remove()

final_grad = model.weight.grad.detach().clone()

print("expected_g1:", expected_g1.cpu().tolist())
print("expected_g2:", expected_g2.cpu().tolist())
print("expected_total:", expected_total.cpu().tolist())
print("bug_total:", bug_total.cpu().tolist())

print("first_grad_ptr:", first_grad_ptr)
print("first_grad_snapshot:", first_grad_snapshot.cpu().tolist())
print("pre_accum_existing_ptr:", observations["existing_ptr"])
print("pre_accum_incoming_ptr:", observations["incoming_ptr"])
print("pre_accum_existing_value:", observations["existing_value"].cpu().tolist())
print("pre_accum_incoming_value:", observations["incoming_value"].cpu().tolist())
print("final_grad:", final_grad.cpu().tolist())

print("first grad == g1:", torch.equal(first_grad_snapshot, expected_g1))
print("pre-accumulate existing grad == g1:", torch.equal(observations["existing_value"], expected_g1))
print("pre-accumulate existing grad == g2:", torch.equal(observations["existing_value"], expected_g2))
print("incoming grad == g2:", torch.equal(observations["incoming_value"], expected_g2))
print("final grad == expected g1+g2:", torch.equal(final_grad, expected_total))
print("final grad == bug g2+g2:", torch.equal(final_grad, bug_total))
```

### Actual output

```text
expected_g1: [3.0, 5.0, 7.0, 9.0]
expected_g2: [11.0, 22.0, 33.0, 44.0]
expected_total: [14.0, 27.0, 40.0, 53.0]
bug_total: [22.0, 44.0, 66.0, 88.0]

first_grad_ptr: 133201995498496
first_grad_snapshot: [3.0, 5.0, 7.0, 9.0]
pre_accum_existing_ptr: 133201995498496
pre_accum_incoming_ptr: 133201995498496
pre_accum_existing_value: [11.0, 22.0, 33.0, 44.0]
pre_accum_incoming_value: [11.0, 22.0, 33.0, 44.0]
final_grad: [22.0, 44.0, 66.0, 88.0]

first grad == g1: True
pre-accumulate existing grad == g1: False
pre-accumulate existing grad == g2: True
incoming grad == g2: True
final grad == expected g1+g2: False
final grad == bug g2+g2: True
```

### Expected behavior

The second backward should accumulate into the first microbatch gradient:

```text
final_grad == [14.0, 27.0, 40.0, 53.0]
```

Instead, the first gradient is overwritten by the second graph replay before accumulation, and the final result is:

```text
final_grad == [22.0, 44.0, 66.0, 88.0]
```

### Environment

Observed on a CUDA-enabled PyTorch build:

```text
PyTorch Version: 2.11.0a0+a6c236b9fd
CUDA available: True
Torch CUDA version: 13.2
```

### Possible fix direction

For module parameter gradient slots, `make_graphed_callables` should not return detached aliases of captured static grad buffers to autograd. Returning an owned tensor, for example cloning parameter grad slots in `Graphed.backward()`, would avoid letting `AccumulateGrad` install a static CUDA graph buffer as `param.grad`.

The current behavior appears inconsistent with the documented drop-in replacement expectation for autograd-enabled training loops.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @mcarilli @eellison @penguinwu @BoyuanFeng

**Comment by eqy** [comment]
Thanks! Currently investigating this with @eee4017 et al.
